internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,606 @@
+		"Error when getting authentication modes": {username: "GAM_error", wantErr: true},
+		"Error when broker returns invalid modes": {username: "GAM_invalid", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			brokerCfgDir := t.TempDir()
+			brokerCfg := testutils.StartBusBrokerMock(t, brokerCfgDir)
+
+			client := startClient(t, brokerCfgDir)
+
+			if tc.brokerID == "" {
+				tc.brokerID = getGeneratedID(t, client, filepath.Base(brokerCfg))
+			}
+
+			if tc.brokerID != "does not exist" {
+				if tc.username == "" {
+					tc.username = "does not matter"
+				}
+				sbResp, err := client.SelectBroker(context.Background(), &authd.SBRequest{
+					BrokerId: tc.brokerID,
+					Username: tc.username,
+				})
+				require.NoError(t, err, "Setup: failed to create session for tests")
+				if tc.sessionID == "" {
+					tc.sessionID = sbResp.GetSessionId()
+				}
+			}
+
+			if tc.layout == nil {
+				tc.layout = normalLayout
+			}
+
+			if tc.sessionID == "empty" {
+				tc.sessionID = ""
+			}
+
+			gamReq := &authd.GAMRequest{
+				SessionId:          tc.sessionID,
+				SupportedUiLayouts: []*authd.UILayout{tc.layout},
+			}
+			gamResp, err := client.GetAuthenticationModes(context.Background(), gamReq)
+			if tc.wantErr {
+				require.Error(t, err, "GetAuthenticationModes should return an error, but did not")
+				return
+			}
+			require.NoError(t, err, "GetAuthenticationModes should not return an error, but did")
+			got := gamResp.GetAuthenticationModes()
+			want := testutils.LoadWithUpdateFromGoldenYAML(t, got)
+			require.Equal(t, want, got, "GetAuthenticationModes returned an unexpected response")
+		})
+	}
+}
+
+func TestSelectAuthenticationMode(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		username  string
+		sessionID string
+		brokerID  string
+		authMode  string
+
+		wantErr bool
+	}{
+		"Successfully select form":        {username: "SAM_form_success"},
+		"Successfully select qrcode":      {username: "SAM_qrcode_success"},
+		"Successfully select newpassword": {username: "SAM_newpassword_success"},
+
+		// service errors
+		"Error when broker does not exist":   {sessionID: "no broker", brokerID: "does not exist", wantErr: true},
+		"Error when sessionID is empty":      {sessionID: "empty", wantErr: true},
+		"Error when no authmode is selected": {sessionID: "no auth mode", authMode: "empty", wantErr: true},
+
+		// broker errors
+		"Error when selecting authentication mode": {username: "SAM_error", wantErr: true},
+
+		/* Layout errors */
+
+		// Layout type errors
+		"Error when returning no layout":           {username: "SAM_no_layout", wantErr: true},
+		"Error when returning invalid layout type": {username: "SAM_invalid_layout_type", wantErr: true},
+
+		// Type "form" errors
+		"Error when returning form w/ no label":      {username: "SAM_form_no_label", wantErr: true},
```

Comment:
mind adding a TODO that should be changed once 23 is merged? (as we don’t hardcode any of those anymore and so, we minimize all those errors to only one with "layout requirements are not fulfilled")

Reply:
The plan was to merge this first and then update the tests on the layouts PR, no?